### PR TITLE
De-duplicate message handling

### PR DESCRIPTION
**Pull request by christianpoveda**
This draft PR aims to de-duplicate the message handling method inside workers. The only blocking issue here is that I haven't been able to desugar the async syntax from the [`process_message`](https://github.com/christianpoveda/bee-p/blob/250ced1c95388a225103bc2f3492c429b7679bb0/bee-protocol/src/worker/peer/handshaker.rs#L222) method in order to  pass its type as an argument [here](https://github.com/christianpoveda/bee-p/blob/250ced1c95388a225103bc2f3492c429b7679bb0/bee-protocol/src/worker/peer/mod.rs#L41).

I think this is the only way of deduplicating that method (other than using a macro as @thibault-martinez suggested as a last resource) that could work without introducing new dependencies to use async traits or something similar.

Edit: I reorganized the code to do all the message handling in an iterator. This iterator yields what `process_message` needs as an input.

### CODE REVIEW

**Comment by christianpoveda** [comment]
@thibault-martinez, this PR is ready for review. I was able to deduplicate some code inside the `run` method of the workers by adding a new `EventHandler` that has a buffer where the bytes of the events are pushed when needed. I also added some tests and documentation.

File: bee-protocol/src/worker/peer/handshaker.rs [hunk 1]
```diff
@@ -116,25 +101,20 @@ impl PeerHandshakerWorker {
...
-                    break;
-                }
+        let mut message_handler = MessageHandler::new(receiver_fused, shutdown_fused, self.peer.address);
+
+        while let Some((header, bytes)) = message_handler.fetch_message().await {
+            if let Err(e) = self.process_message(&header, bytes).await {
+                error!("[{}] Processing message failed: {:?}.", self.peer.address, e);
+            }
+            if let HandshakeStatus::Done | HandshakeStatus::Duplicate = self.status {
+                break;
```
Comment (christianpoveda) [inline bee-protocol/src/worker/peer/handshaker.rs:111]: @thibault-martinez here

File: bee-protocol/src/worker/peer/message_handler.rs [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+/// This type is used by `MessageHandler` to decide what should be read next when handling an
+/// event.
+enum ReadState {
+    /// `MessageHandler` should read a header.
+    Header,
+    /// `MessageHandler` should read a payload based on a header.
+    Payload(Header),
+}
+
+/// A message handler
```
Comment (thibault-martinez) [inline bee-protocol/src/worker/peer/message_handler.rs:38]:
```suggestion
/// A message handler.
```

File: bee-protocol/src/worker/peer/message_handler.rs [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    }
+    /// Fetch the header and payload of a message.
+    pub(super) async fn fetch_message<'a>(&'a mut self) -> Option<(Header, &'a [u8])> {
+        // loop until we can return the header and payload
+        loop {
+            match &self.state {
+                // Read a header.
+                ReadState::Header => {
+                    // We need 3 bytes to read a header.
+                    let bytes = self.events.fetch_bytes(3).await?;
```
Comment (thibault-martinez) [inline bee-protocol/src/worker/peer/message_handler.rs:67]: Maybe use the constant `HEADER_SIZE` instead of 3

Comment (christianpoveda) [reply bee-protocol/src/worker/peer/message_handler.rs:67]: sure!

**Pull request merged**